### ISSUE TITLE
Storybook is not able to resolve path with Next js absolute import

### ISSUE THREAD

**Issue by rahul3103** [storybookjs/storybook#11639]
**Describe the bug**
Storybook is not able to resolve path with Next js absolute import

**To Reproduce**
yarn add storybook
Added .storybook folder
Add jsconfig for absolute import

**Expected behavior**
It should build correctly.
```
const path = require('path');
const resolve = (dir) => path.resolve(__dirname, dir);

module.exports = ({ config }) => {
  config.resolve.alias = {
    '@utils/*': resolve('../src/utils/*'),
   ....
  };

  return config;
};

```

**Error**
```

ERROR in ./src/components/Base/Button/Button.jsx
Module not found: Error: Can't resolve '@utils/boxShadowGenerator' in '/Users/rahulshrivastava/projects/unacademy-web-learning/src/components/Base/Button'
 @ ./src/components/Base/Button/Button.jsx 9:0-59 107:35-53 122:9-27
 @ ./src/stories/button.stories.js
 @ ./src/stories sync \.stories\.js$
 @ ./.storybook/config.js
 @ multi ./node_modules/@storybook/core/dist/server/common/polyfills.js ./node_modules/@storybook/core/dist/server/preview/globals.js ./.storybook/config.js (webpack)-hot-middleware/client.js?reload=true&quiet=true
```

**Code snippets**
I am using jsconfig.json
```
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es2016",
    "jsx": "react",
    "baseUrl": "./src/",
    "paths": {
      "@utils/*": ["utils/*"],
    ....
    }
  },
  "exclude": ["node_modules", "dist"]
}
```

**System:**
```
 npx -p @storybook/cli@next sb info

Environment Info:
(node:1828) UnhandledPromiseRejectionWarning: TypeError: e.filter is not a function
    at /Users/rahulshrivastava/.npm/_npx/1031/lib/node_modules/@storybook/cli/node_modules/envinfo/dist/envinfo.js:1:73205
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Promise.all (index 6)
(node:1828) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:1828) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by EduardoHidalgo** [comment]
I have exactly the same problem! I tried adding a tsconfig.json and with a custom webpack but didn't work.

**Comment by altechzilla** [comment]
+1 - Same issue here. Any idea on how to process Next.js paths in jsconfig.json? I found this on TypeScript (haven't tested it) but need one on Javascript instead: https://github.com/storybookjs/storybook/issues/9610

**Comment by kalanda** [comment]
+1 Same problem here. NextJS resolves but SB don't.

jsconfig.json
```
{
  "compilerOptions": {
    "baseUrl": "./src"
  }
}
```

**Comment by EduardoHidalgo** [comment]
@kalanda That's correct, Next.js resolve but SB don't. I tried a lot of webpack solutions and multiple tsconfig.json options but anyone was successful. I decided to remove SB for my project for now. Sad.

**Comment by GxDesign** [comment]
I havent gotten everything working with Next yet but this resolved the absolute import issues: 
`const path = require('path');

module.exports = ({config}) => {
  config.resolve.modules = [
    path.resolve(__dirname, ".."),
    "node_modules",
  ]

  return config
}
`

For using ./src,  you can use path.resolve(__dirname, "..", "src"),

**Comment by EduardoHidalgo** [comment]
With SB 6.x you don't have a webpack configuration, instead you use a main.js file with specific import structure (see [Storybook React Configure Preview](https://storybook.js.org/docs/react/configure/overview)). Please, could you provide the same code in this format?

> I havent gotten everything working with Next yet but this resolved the absolute import issues (in .storybook/webpack.config.js):
> 
> ```
> const path = require('path');
> 
> module.exports = ({config}) => {
>   config.resolve.modules = [
>     path.resolve(__dirname, ".."),
>     "node_modules",
>   ]
> 
>   return config
> }
> ```
> 
> For using ./src, you can use path.resolve(__dirname, "..", "src"),

**Comment by GxDesign** [comment]
You can still add a webpack.config.js to your .storybook directory. But with main.js, you define it as webpackFinal: 

```
module.exports = {
  stories: ["../stories/*.stories.@(js|jsx)"],
  addons: [
    "@storybook/addon-actions",
    "@storybook/addon-knobs",
    "@storybook/addon-links",
  ],
  presets: ["@storybook/preset-scss"],
  webpackFinal: async (config, { configType }) => {
    config.resolve.modules = [
      path.resolve(__dirname, ".."),
      "node_modules",
    ]

    return config;
  }
};
```

**Comment by Fernando-FloresP** [comment]
Hello! I've just tried this fix in a project with Nextjs v9.5.3 and Storybook v6.0.21 but its not working. 

On the path /sb-test/stories/Button.stories.js on the the 2nd and 3rd lines you can switch between absolute and relative paths.
The absolute path works but if i use the relative path it doesn't work. 

> //import { ButtonComponent } from "../src/components/button/Button";
> import { ButtonComponent } from "@/Components/button/Button";

This is the error:
![image](https://user-images.githubusercontent.com/69826654/92794087-cf6cee80-f374-11ea-9c57-8158a1151f6c.png)

>ERROR in ./stories/Button.stories.js
Module not found: Error: Can't resolve '@/Components/button/Button' in '/mnt/c/Users/reape/sb-test/stories'
 @ ./stories/Button.stories.js 85:0-61 88:13-28 97:42-57
 @ ./stories sync ^\.(?:(?:^|\/|(?:(?:(?!(?:^|\/)\.).)*?)\/)(?!\.)(?=.)[^/]*?\.stories\.(js|jsx|ts|tsx))$
 @ ./.storybook/generated-stories-entry.js
 @ multi ./node_modules/@storybook/core/dist/server/common/polyfills.js ./node_modules/@storybook/core/dist/server/preview/globals.js ./.storybook/storybook-init-framework-entry.js ./node_modules/@storybook/addon-docs/dist/frameworks/common/config.js-generated-other-entry.js ./node_modules/@storybook/addon-docs/dist/frameworks/react/config.js-generated-other-entry.js ./node_modules/@storybook/addon-links/dist/preset/addDecorator.js-generated-other-entry.js ./node_modules/@storybook/addon-actions/dist/preset/addDecorator.js-generated-other-entry.js ./node_modules/@storybook/addon-actions/dist/preset/addArgs.js-generated-other-entry.js ./node_modules/@storybook/addon-backgrounds/dist/preset/defaultParameters.js-generated-other-entry.js ./.storybook/preview.js-generated-config-entry.js ./.storybook/generated-stories-entry.js (webpack)-hot-middleware/client.js?reload=true&quiet=false&noInfo=undefined

**Comment by GxDesign** [comment]
Its because thats not an absolute import from root, your path is using an alias. The following code allows you to import from src using "src/components/button/Button": 
```
config.resolve.modules = [
      path.resolve(__dirname, ".."),
      "node_modules",
    ]
```

But it does not tell webpack anything about the "@/Components" syntax. You would need to pass an alias configuration like this: 

```
config.resolve.alias = {
      ...config.resolve.alias,
      "@/Components": path.resolve(__dirname, "../src/components")
    };
```

**Comment by Fernando-FloresP** [comment]
Great! It worked perfectly, thanks a lot man. I wrote a commit implementing your fix: [Absolute path with Storybook problem solved](https://github.com/Fernando-FloresP/sb-test/commit/6162d38d29bd3e90261946eb8e9423ef8ac87e57)

**Comment by EduardoHidalgo** [comment]
I love you man, I spent like about 40 hours trying to solve but I never could. Thanks <3

> Its because thats not an absolute import from root, your path is using an alias. The following code allows you to import from src using "src/components/button/Button":
> 
> ```
> config.resolve.modules = [
>       path.resolve(__dirname, ".."),
>       "node_modules",
>     ]
> ```
> 
> But it does not tell webpack anything about the "@/Components" syntax. You would need to pass an alias configuration like this:
> 
> ```
> config.resolve.alias = {
>       ...config.resolve.alias,
>       "@/Components": path.resolve(__dirname, "../src/components")
>     };
> ```

**Comment by GxDesign** [comment]
@EduardoHidalgo haha no problem, glad I could help 🍻

**Comment by bananashell** [comment]
Is this the preferred way to solve this, should perhaps be added to the docs if it is?

**Comment by shilman** [comment]
Hey @eric-burel do you mind taking a look at this one?

**Comment by eric-burel** [comment]
Basically the underlying issue, is that you need to reuse the same Webpack config in both Next and Storybook, and make sure your Webpack config is robust enough to be reused, eg using `__dirname`

For example I use this in Vulcan:
```js
const withMagicImports = (config = {}) => {
  if (!config.resolve) config.resolve = {};
  config.resolve.alias = {
    ...config.resolve.alias,
    "~": path.join(__dirname, "../../../", "src"),
  };
  config.resolve.modules = [
    ...(config.resolve.modules || []),
    path.join(__dirname, "../../../", "packages"),
  ];
  return config;
};
```
I use that both in `next.config.js` and `main.js` to extend the Webpack config. It is robust to path changes because I use `__dirname`.

Maybe I can add a section in https://storybook.js.org/docs/react/configure/webpack for this advanced usage with a code sample if that's a pretty common issue?

**Comment by eric-burel** [comment]
I've built a repro: https://github.com/lbke/next-sb-default-absolute-imports

I'll try to see if we can make it work out of the box in Storybook by using the same pattern as Next. 

In the meantime the official solution is to get the Storybook config right in order to have the same behavior as Next, as demonstrated in the comment above.

**Comment by eric-burel** [comment]
@shilman Everything is happening in this file in Next `packages/next/build/webpack-config.ts` as far as I can tell. It will parse `tsconfig` and `jsconfig` to explicitely get the alias. I am asking Next people to get confirmation. Could it be worthy to add this in Storybook also? I'll try to demo that but it'll take a bit of time.

**Comment by eric-burel** [comment]
We might want to check this experimental plugin from Next: https://github.com/vercel/next.js/tree/canary/packages/next-plugin-storybook, it makes use of Next webpack config

**Comment by jonniebigodes** [comment]
@eric-burel it would be fantastic if the plugin was updated and allowed Storybook 6.x to be integrated more smoothly. So far i've tested it in a dummy repo and once you run `yarn storybook` it yields the following:

![next-plugin-error](https://user-images.githubusercontent.com/22988955/97057474-32ac8c00-1583-11eb-8e5f-354ac17ce1bc.png)

Based on a preliminary search it seems that the last change done to the plugin was pre 6.x release date, around 5/6 months ago.

I'm going to raise an issue asking if the plugin is going to be updated to support Storybook 6.0 as it would help out not only Storybook users and NextJs users as well.

**Comment by eric-burel** [comment]
See https://github.com/vercel/next.js/pull/18364
When fixed the preset works perfect, I'll PR a TypeScript version in Next to make it more robust

**Comment by eric-burel** [comment]
Even better: https://github.com/vercel/next.js/pull/18367
If someone is good at TypeScript, giving some help to fix `packages/next-plugin-storybook/tsconfig.json` to finish the PR would be awesome. This would make the plugin more reliable.

**Comment by gpessa** [comment]
Di we have any news about this one?

**Comment by eric-burel** [comment]
Current direction would be at Next.js level, splitting Next webpack logic feature by feature. It is currently quite monolithic and does a lot of things, which in turns makes it quite difficult to stabilize this plugin. I am personnaly ok to invest time in this but I might be busy for a while :/ so contributions are welcome. 

Taking a look at https://github.com/vercel/next.js/pull/18367/files could help. 

Also you may want to check https://github.com/UnlyEd/next-right-now/pull/251, Storybook has been integrated in Next Right Now with good support. I think solving the issue more elegantly at Next level could help simplifiying the code but it seems to do the job.

**Comment by EduardoHidalgo** [comment]
> Current direction would be at Next.js level, splitting Next webpack logic feature by feature. It is currently quite monolithic and does a lot of things, which in turns makes it quite difficult to stabilize the Storybook plugin. I am personnaly ok to invest time in this but I might be busy for a while :/ so contributions are welcome.
> 
> Taking a look at https://github.com/vercel/next.js/pull/18367/files could help.
> 
> Also you may want to check [UnlyEd/next-right-now#251](https://github.com/UnlyEd/next-right-now/pull/251), Storybook has been integrated in Next Right Now with good support. I think solving the issue more elegantly at Next level could help simplifiying the code but it seems to do the job.

I don't think NRN should be one the first option. Personally I always prefer an unopininated option. Some SB plugin or an internal Next.js solution will be better.

If you are nice in guide me through the process, I can make a PR (:

**Comment by eric-burel** [comment]
As a starting point you can take a look at this issue, on Next side: https://github.com/vercel/next.js/issues/19345
Relevant logic lives here: https://github.com/vercel/next.js/blob/canary/packages/next/build/webpack-config.ts. You can see that Next has some logic to alter the Webpack config, adding aliases, based on the content of `tsconfig` or `jsconfig` files. 

You'd want to extract this logic to make it more reusable, that could be the starting point for a simpler Next/storybook plugin. Then you can do this process again and again for each part of the config: going step by step would allow to detect and correct the hard part of each feature (loading CSS, using path aliases, etc. etc.).

What I meant for NRN is that you can take inspiration from their Storybook config without actually using it if you don't feel like it.

**Comment by Ahmed-osama** [comment]
> Its because thats not an absolute import from root, your path is using an alias. The following code allows you to import from src using "src/components/button/Button":
> 
> ```
> config.resolve.modules = [
>       path.resolve(__dirname, ".."),
>       "node_modules",
>     ]
> ```
> 
> But it does not tell webpack anything about the "@/Components" syntax. You would need to pass an alias configuration like this:
> 
> ```
> config.resolve.alias = {
>       ...config.resolve.alias,
>       "@/Components": path.resolve(__dirname, "../src/components")
>     };
> ```

thank you man you saved my day 

that's how it worked for **next.js**

```js
const path = require('path');
module.exports = {
  stories: [
    '../desktop/**/*.stories.@(ts|tsx)',
    '../mobile/**/*.stories.@(ts|tsx)',
  ],
  addons: ['@storybook/addon-links', '@storybook/addon-essentials'],
  webpackFinal: async (config, { configType }) => {
    config.resolve.modules = [path.resolve(__dirname, '..'), 'node_modules'];
    config.resolve.alias = {
      ...config.resolve.alias,
      '@': path.resolve(__dirname, '../'),
    };
    return config;
  },
};

```

**Comment by eric-burel** [comment]
Hi, I am going to close this for now, as there are 2 main alternatives:

- just set Storybook up to reproduce Next.js magic, there are multiple comments demoing how to do so. This is of course not very satisfactory, but it works for the moment.

- in the meantime we will keep working on `@next/plugin-storybook` to make this a one-liner, but this takes time as it is not so easy to export Next.js buildtime magic to 3rd parties at the moment. It technically works, but it's not very stable. I'll demo this package usage here: https://github.com/lbke/next-plugin-storybook-demo. It may be included in Next.js as an official example but only when `@next/plugin-storybook` reach stability. Anyway, this question is on Next.js side at the moment more than SB.

**Comment by Matheus8174** [comment]
now just install the next addon in your project.

https://storybook.js.org/addons/storybook-addon-next

**Comment by nabu1615** [comment]
After some errors and look a way to work with NEXT components this code works for me.

`webpackFinal: async (config, { configType }) => {
        config.resolve.modules = [
            path.resolve(__dirname, '..'),
            'node_modules',
        ];

        config.resolve.alias = {
            ...config.resolve.alias,
            components: path.resolve(__dirname, '../components'),
        };

        return config;
    },`

**Comment by sendy34** [comment]
Anyone still having issues, I had two typescript configurations.
One was on global `ts-config` but there was another configuration on `main.ts` for Storybook.
If you have a custom configuration on `main.ts`, then `alias` also needs to be defined on that configuration under `paths`.

Once you define the alias, and resolve the alias using `webpackFinal`, it should start working fine!

**Comment by shilman** [comment]
@valentinpalkovic @vanessayuenn This was raised [on Twitter](https://x.com/sprog_ch/status/1768408725633511534?s=20) I've reproduced by editing the Storybook CLI template files:

```tsx
// stories/Header.tsx
import React from 'react';
import { Button } from '@/stories/Button';
```

`@/*` is installed in the Next.js boilerplate by default, so I think we should prioritize fixing this in `@storybook/nextjs`. WDYT?

**Comment by valentinpalkovic** [comment]
I would love to work on it!

**Comment by gabitoesmiapodo** [comment]
@shilman @valentinpalkovic I don't know if it helps you or not, but I got it working by doing this: https://stackoverflow.com/a/68288720/7949258

In any case it might help someone having this problem (like me!).

**Comment by weyert** [comment]
I thought Storybook already used `tsconfig-paths-webpack-plugin`?

**Comment by valentinpalkovic** [comment]
Actually, `tsconfig-paths-webpack-plugin` is configured for Next.js. You have to define the `baseUrl` in your tsconfig.json, though, to activate the plugin. If `baseUrl` isn't defined, the plugin doesn't get initialized.

**Comment by thewanionly** [comment]
For my case, adding `"baseUrl": "."` in my `tsconfig.json` fixes the issue:

```
    "baseUrl": ".",
    "paths": {
      "@/*": ["./src/*"]
    }
```

**Comment by valentinpalkovic** [comment]
This was fixed by https://github.com/storybookjs/storybook/pull/26651. The fix will be part of the 8.0.6 release!
